File: python-legacy/orchestrator/bss_orchestrator/safety.py
"""Destructive-operation gating for LLM tool calls.

Policy: some tools can't be undone with a follow-up call (terminate a
subscription, remove a payment method, cancel an order). If the human hasn't
explicitly passed ``--allow-destructive``, these tools return a *structured*
error rather than executing:

    {"error": "DESTRUCTIVE_OPERATION_BLOCKED", "tool": "...", "message": "..."}

The LangGraph supervisor sees this structured error as the tool result and
either aborts cleanly or asks the user to re-run with the flag. Same pattern
as ``PolicyViolationFromServer`` — structured error, never a stack trace.

v1.5 — autonomy-aware gating. The base contract above is unchanged:
``allow_destructive=False`` always blocks. What v1.5 adds is differentiated
behaviour when ``allow_destructive=True``:

- ``autonomy_mode="batched"`` — the historical behaviour. Once allowed,
  every destructive tool call in the current graph instance executes.
  The cockpit's per-turn ``allow_destructive=True`` flag thus authorises
  the whole loop.
- ``autonomy_mode="granular"`` (the new default) — only the FIRST destructive
  call in the current graph instance executes; subsequent destructive calls
  block again with the same structured error, so the LLM must re-propose
  and the operator must ``/confirm`` for each one. Per-step operator
  control even after the first authorisation.

The per-graph "first-destructive-fired" state lives in a ``LoopState``
dict that ``build_tools`` creates fresh and shares across every wrapped
destructive tool in the same ``build_graph`` invocation. Each ``astream_once``
call builds its own graph, so the state naturally resets between turns —
no manual reset required. Test paths that wrap a single tool directly
get an implicit private LoopState.

The destructive-tool list (``DESTRUCTIVE_TOOLS``) is unchanged in v1.5.
Autonomy controls *how many* ``/confirm``s a compound action needs, NOT
*which* tools require one. Adding a tool to ``DESTRUCTIVE_TOOLS`` is still
a doctrine decision that requires reviewing the safety contract.
"""

from __future__ import annotations

import functools
from collections.abc import Awaitable, Callable
from typing import Any

# Every destructive tool in the registry. Matches the dotted name the LLM sees
# (``<domain>.<action>``), NOT the Python function name — LangGraph tools are
# registered with the dotted name via ``@tool(name=...)``.
DESTRUCTIVE_TOOLS: frozenset[str] = frozenset(
    {
        "customer.close",
        "customer.remove_contact_medium",
        "case.close",
        "ticket.cancel",
        "payment.remove_method",
        "order.cancel",
        "subscription.terminate",
        # v0.12 — chat-surface wrapper around subscription.terminate.
        # The wrapper narrows the LLM-visible target to the actor's
        # own line, but the operation is still irreversible — gating
        # at the wrapper too keeps the destructive contract honest
        # even if a future caller forgets allow_destructive=True.
        "subscription.terminate_mine",
        "provisioning.set_fault_injection",
        "admin.reset_operational_data",
        "admin.force_state",
    }
)


# v1.5 — per-graph-build mutable state shared across destructive wrappers.
# A plain dict keeps the contract obvious from the wrapper's call site.
# Single key today; reserving the dict shape so v1.5.x (per-tool autonomy
# annotations, per-session overrides) can extend it without a signature
# change ripple.
LoopState = dict[str, Any]


def make_loop_state() -> LoopState:
    """Return a fresh per-graph mutable state for autonomy-aware gating.

    Called once by ``build_tools`` per ``build_graph`` invocation; shared
    across every destructive wrapper in that graph so they observe a
    consistent "has any destructive fired in this loop yet?" answer.
    """
    return {"destructive_executed": 0}


def is_destructive(tool_name: str) -> bool:
    """True if ``tool_name`` (dotted, e.g. ``subscription.terminate``) is gated."""
    return tool_name in DESTRUCTIVE_TOOLS


def _blocked_response(tool_name: str) -> dict[str, Any]:
    return {
        "error": "DESTRUCTIVE_OPERATION_BLOCKED",
        "tool": tool_name,
        "message": (
            f"Tool {tool_name!r} is destructive and requires operator "
            "/confirm. Propose this tool to the operator by stopping after "
            "your proposal; the next /confirm-bracketed turn will execute "
            "it."
        ),
    }


def wrap_destructive(
    tool_fn: Callable[..., Awaitable[Any]],
    *,
    tool_name: str,
    allow_destructive: bool,
    autonomy_mode: str = "batched",
    loop_state: LoopState | None = None,
) -> Callable[..., Awaitable[Any]]:
    """Return a coroutine wrapper that short-circuits if the tool is
    destructive and the gate is closed.

    Non-destructive tools are returned unchanged — no overhead.

    Args:
        tool_fn: the underlying tool coroutine.
        tool_name: dotted LLM-facing name (``"subscription.terminate"``).
        allow_destructive: master gate set by the caller (the CLI's
            ``--allow-destructive`` or the cockpit's per-turn
            ``allow_destructive=True`` after ``/confirm``).
        autonomy_mode: ``"granular"`` (re-gate after first destructive
            fires) or ``"batched"`` (one authorisation covers the
            loop). Defaults to ``"batched"`` to preserve pre-v1.5
            behaviour for callers that haven't been updated yet
            (mostly test paths); production call sites in v1.5+ pass
            the autonomy mode explicitly from
            ``read_autonomy_mode()`` cached on app/REPL state.
        loop_state: per-graph state created by ``make_loop_state()``;
            shared across every destructive wrapper in the same
            ``build_graph`` invocation. When ``None``, a private
            state is created (test-path convenience — production
            paths via ``build_tools`` always pass a shared state).
    """
    if not is_destructive(tool_name):
        return tool_fn

    state = loop_state if loop_state is not None else make_loop_state()

    @functools.wraps(tool_fn)
    async def _gated(**kwargs: Any) -> Any:
        if not allow_destructive:
            return _blocked_response(tool_name)
        # allow_destructive=True path.
        # Granular mode re-gates after the first destructive in this
        # graph has executed; batched mode authorises the whole loop.
        if (
            autonomy_mode == "granular"
            and state["destructive_executed"] >= 1
        ):
            return _blocked_response(tool_name)
        state["destructive_executed"] += 1
        return await tool_fn(**kwargs)

    return _gated

File: rust/crates/bss-orchestrator/src/safety.rs
//! Destructive-operation gating for LLM tool calls. Port of
//! `orchestrator/bss_orchestrator/safety.py`.
//!
//! Some tools can't be undone with a follow-up call. If the human hasn't passed
//! `allow_destructive`, they return a *structured* error rather than executing:
//!
//! ```json
//! {"error": "DESTRUCTIVE_OPERATION_BLOCKED", "tool": "...", "message": "..."}
//! ```
//!
//! v1.5 autonomy: when `allow_destructive=true`, `batched` authorises the whole
//! loop after the first destructive fires; `granular` (the cockpit default)
//! re-gates each destructive so the operator must `/confirm` for each one.

use serde_json::{json, Value};

/// Every destructive tool in the registry (dotted LLM-facing name). Adding a name
/// here is a doctrine decision (requires reviewing the safety contract).
pub const DESTRUCTIVE_TOOLS: &[&str] = &[
    "customer.close",
    "customer.remove_contact_medium",
    "case.close",
    "ticket.cancel",
    "payment.remove_method",
    "order.cancel",
    "subscription.terminate",
    // v0.12 — chat-surface wrapper around subscription.terminate. Gating at the
    // wrapper too keeps the destructive contract honest.
    "subscription.terminate_mine",
    "provisioning.set_fault_injection",
    "admin.reset_operational_data",
    "admin.force_state",
];

/// Autonomy mode for destructive gating when `allow_destructive=true`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AutonomyMode {
    /// One authorisation covers the whole loop (pre-v1.5 behaviour).
    Batched,
    /// Re-gate each destructive after the first fires (cockpit default).
    Granular,
}

/// True if `tool_name` (dotted) is gated.
pub fn is_destructive(tool_name: &str) -> bool {
    DESTRUCTIVE_TOOLS.contains(&tool_name)
}

/// Per-graph mutable state shared across destructive gating in one loop so
/// granular mode can observe "has any destructive fired yet?".
#[derive(Debug, Default)]
pub struct LoopState {
    pub destructive_executed: u32,
}

/// The structured block the LLM sees when a destructive tool is refused.
pub fn blocked_response(tool_name: &str) -> Value {
    json!({
        "error": "DESTRUCTIVE_OPERATION_BLOCKED",
        "tool": tool_name,
        "message": format!(
            "Tool {tool_name:?} is destructive and requires operator /confirm. \
             Propose this tool to the operator by stopping after your proposal; \
             the next /confirm-bracketed turn will execute it."
        ),
    })
}

/// Decide whether a destructive tool call may execute this turn, mutating
/// `state` to record a fire. Non-destructive tools always pass. Returns `Err`
/// carrying the structured block when refused.
///
/// Mirrors `wrap_destructive`'s runtime check: `allow_destructive=false` always
/// blocks; in granular mode a second destructive re-blocks.
pub fn gate_destructive(
    tool_name: &str,
    allow_destructive: bool,
    autonomy: AutonomyMode,
    state: &mut LoopState,
) -> Result<(), Value> {
    if !is_destructive(tool_name) {
        return Ok(());
    }
    if !allow_destructive {
        return Err(blocked_response(tool_name));
    }
    if autonomy == AutonomyMode::Granular && state.destructive_executed >= 1 {
        return Err(blocked_response(tool_name));
    }
    state.destructive_executed += 1;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn non_destructive_always_passes() {
        let mut st = LoopState::default();
        assert!(gate_destructive("customer.get", false, AutonomyMode::Granular, &mut st).is_ok());
    }

    #[test]
    fn blocked_when_not_allowed() {
        let mut st = LoopState::default();
        let r = gate_destructive(
            "subscription.terminate",
            false,
            AutonomyMode::Batched,
            &mut st,
        );
        assert!(r.is_err());
        assert_eq!(st.destructive_executed, 0);
    }

    #[test]
    fn batched_authorises_the_whole_loop() {
        let mut st = LoopState::default();
        // Every destructive fires once allowed.
        assert!(gate_destructive(
            "subscription.terminate",
            true,
            AutonomyMode::Batched,
            &mut st
        )
        .is_ok());
        assert!(gate_destructive("order.cancel", true, AutonomyMode::Batched, &mut st).is_ok());
        assert_eq!(st.destructive_executed, 2);
    }

    #[test]
    fn granular_regates_after_first_fire() {
        let mut st = LoopState::default();
        // First destructive passes; the second re-blocks until a fresh /confirm.
        assert!(gate_destructive(
            "subscription.terminate",
            true,
            AutonomyMode::Granular,
            &mut st
        )
        .is_ok());
        assert!(gate_destructive("order.cancel", true, AutonomyMode::Granular, &mut st).is_err());
        assert_eq!(st.destructive_executed, 1);
    }
}
